lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -1229,7 +1275,37 @@ func (d *zfs) RefreshVolume(vol Volume, srcVol Volume, srcSnapshots []Volume, al
 
 // DeleteVolume deletes a volume of the storage device. If any snapshots of the volume remain then
 // this function will return an error.
+// For image volumes, both filesystem and block volumes will be removed.
 func (d *zfs) DeleteVolume(vol Volume, op *operations.Operation) error {
+	for _, suffix := range blockBackedAllowedFilesystems {
+		tmpVol := vol
+
+		// We need to deep copy the config otherwise changing `zfs.block_mode`
+		// in tmpVol will also change it in vol.
+		tmpVol.config = map[string]string{}
+		tmpVol.poolConfig = map[string]string{}
+
+		for k, v := range vol.config {
+			tmpVol.config[k] = v
+		}
+
+		for k, v := range vol.poolConfig {
+			tmpVol.config[k] = v
+		}
+
+		tmpVol.config["zfs.block_mode"] = "false"
```

Comment:
This would be clearer if you left `zfs.block_mode` alone (i.e "true) and instead changed `block.filesystem` to the loop filesystem. Then the logic in `dataset()` line 41 should kick in an produce the correct volume name without having to modify the volume name manually here.